docs/storage/upload/upload.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2017 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	"google.golang.org/api/storage/v1"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+)
+
+func main() {
+	// Prevent log from printing out time information
+	log.SetFlags(0)
+
+	var projectID, bucket, source, name string
+	var public bool
+
+	flag.StringVar(&bucket, "bucket", "test", "the bucket to upload content to")
+	flag.StringVar(&projectID, "project-id", "golang-samples", "the ID of the GCP project to use")
+	flag.StringVar(&source, "source", "", "the path to the source")
+	flag.StringVar(&name, "name", "", "the name of the file to be stored on GCS")
+	flag.BoolVar(&public, "public", true, "whether the item should be available publicly")
+	flag.Parse()
+
+	storageSrvc, err := DefaultGCSService()
+	if err != nil {
+		log.Fatalf("creating storage service: %v", err)
+	}
+
+	var r io.Reader
+	if source == "" {
+		r = os.Stdin
+		log.Printf("Reading from stdin...")
+	} else {
+		f, err := os.Open(source)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer f.Close()
+
+		r = f
+	}
+
+	if name == "" {
+		if source != "" {
+			name = filepath.Base(source)
+		} else {
+			name = "test-sample"
```

Comment:
What do you mean by `a timestamped name, e.g. stdin?`, do you mean e.g. `fmt.Sprintf("stdin-%d", rand.Int())`?